### PR TITLE
Testing changes for AES-GCM backed by ALT implementation

### PR DESCRIPTION
**Pull request by stevew817**
## Description
* Allow an AES-GCM ALT implementation to return 'not supported' when it doesn't support non-12-byte IVs
* Add broader AES-GCM vectors (varying key sizes, tag lengths, AAD lengths, ...) to the PSA Crypto suite
* Allow AES-192 based AES-GCM vectors to be skipped if there is an alternative implementation underneath (which could return 'not supported' for AES-192)
* Allow non-12-byte IV AES-GCM vectors to be skipped if there is an alternative implementation underneath (which could return 'not supported' for non-12-byte IVs)
* Set default IV length for 'check that we can exercise an AES-GCM key) to the standard 12 byte IV

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

### CODE REVIEW

File: tests/suites/test_suite_psa_crypto.function
Comment (ronald-cron-arm) [inline tests/suites/test_suite_psa_crypto.function:4089]: Well, this doesn't seem a scalable approach to me and I don't feel I can approve that. We have to find a better solution for the problem of the dependencies of tests going beyond the configuration options: key size, nonce size ...

File: tests/suites/test_suite_psa_crypto.function
Comment (ronald-cron-arm) [inline tests/suites/test_suite_psa_crypto.function:4100]: ditto, I don't feel I can approve that.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
It's very good to have additional AES related tests in the PSA crypto test suite and thanks for that. I have blocking concerns about other parts of the PR though.

File: library/gcm.c
Comment (stevew817) [reply library/gcm.c:825]: No, it's not anecdotal. If you look at the GCM spec, there are two distinct flows based on IV size: 12-byte and non-12-byte.

A 12-byte IV is used directly, whereas a non-12-byte IV needs to be converted to an input IV first using the GHASH primitive. That's an extra processing step beyond having a 'plain' 12-byte IV. Non-12-byte IVs are therefore less common, and thus not all accelerators include support for it.

File: tests/suites/test_suite_psa_crypto.function
Comment (stevew817) [reply tests/suites/test_suite_psa_crypto.function:4089]: I don't think there's going to be a scalable solution before we get the driver JSON autogeneration done, and thus can autogenerate defines based on what's actually accelerator-supported in terms of key sizes. Nonce size I don't think is included in the driver JSON.

This PR is what we have had to do to enable testing on our targets. I agree that it is food for thought for a future improvement (based on the driver JSON), but for now, we'd highly appreciate getting this in. I feel the exception case for skipping is so well-constrained that it in practice won't have impact on anything else besides on-target testing with accelerators present (which you guys don't do yet?).

File: tests/suites/test_suite_psa_crypto.function
Comment (stevew817) [reply tests/suites/test_suite_psa_crypto.function:4100]: See my other comment.

File: tests/suites/test_suite_psa_crypto.function
Comment (stevew817) [reply tests/suites/test_suite_psa_crypto.function:495]: I couldn't find any. Is there one?

File: library/gcm.c
Comment (stevew817) [reply library/gcm.c:819]: fixed

File: tests/suites/test_suite_psa_crypto.data
Comment (stevew817) [reply tests/suites/test_suite_psa_crypto.data:1600]: fixed

File: tests/suites/test_suite_psa_crypto.function
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_psa_crypto.function:4089]: Even with driver descriptions, there will always be cases of partial support that can't be expressed in the description language. So I think the logic here is the best we can do (possibly wrapped in some macros/functions if applicable): try the operation, require either success or not-supported, and the built-in implementation must succeed but drivers/ALT are allowed to return not-supported.

File: tests/suites/test_suite_psa_crypto.function
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_psa_crypto.function:495]: See the implementation of `psa_aead_setup`:
```
if( PSA_ALG_AEAD_WITH_TAG_LENGTH( alg, 0 ) == PSA_ALG_AEAD_WITH_TAG_LENGTH( PSA_ALG_GCM, 0 ) )
```
or (a little more readable but less efficient because it has to do case analysis on the algorithm):
```
if( PSA_ALG_AEAD_WITH_DEFAULT_TAG_LENGTH( alg ) == PSA_ALG_AEAD_WITH_DEFAULT_TAG_LENGTH( PSA_ALG_GCM ) )
```
I admit it doesn't exactly roll off the tongue. Should we add `PSA_ALG_IS_GCM`? Or a more general `PSA_ALG_IS_SAME_AEAD`? Mind you, when you have a switch rather than an if, the existing macros are about as good as it gets.

Comment (stevew817) [reply tests/suites/test_suite_psa_crypto.function]: Thanks for the suggestion. Updated.

**Comment by stevew817** [comment]
@ronald-cron-arm Given our discussion on Monday, are you happy with the PR now?

**Comment by ronald-cron-arm** [comment]
I think there is an action on me for this one to define some "AES-192, GCM nonce-length !=12 ... macros".

**Review by tom-daubney-arm — DISMISSED** [review]
Pending @ronald-cron-arm's modifications to add additional macros to make the code a bit cleaner, I am happy that this PR can be approved.

**Comment by stevew817** [comment]
@ronald-cron-arm Care to point me in the direction of where/how you want those macros? I'd like to progress this PR.

File: tests/suites/test_suite_psa_crypto.function
Comment (ronald-cron-arm) [reply tests/suites/test_suite_psa_crypto.function:4092]: Looking at this, I am wondering if we want to support the checks on MBEDTLS_AES_ALT and MBEDTLS_AES_SETKEY_ENC_ALT ? I guess this is for the case where the AES accelerator is leveraged through the Mbed TLS AES implementation. Don't we want to avoid that and that an AES accelerator is only leveraged through the PSA driver interface. @stevew817, is it currently an issue for you if we remove  #if defined(MBEDTLS_AES_ALT) ||  defined(MBEDTLS_AES_SETKEY_ENC_ALT) here?

Comment (gilles-peskine-arm) [reply tests/suites/test_suite_psa_crypto.function:4092]: It does make sense to use ALT interfaces for accelerators where they exist, since that automatically makes them available under both PSA and non-PSA APIs. PSA is the future, but the present includes ALT as well.

Comment (stevew817) [reply tests/suites/test_suite_psa_crypto.function:4092]: Yes, that would be an issue for us. We're not creating PSA drivers for our 'older' parts (which only have an AES accelerator), and instead are continuing to rely on the mbed TLS implementation and accelerating the primitive underneath there. That's the best we can do ATM to avoid re-implementing a bunch of sensitive software algorithms in a patchwork driver.

Not having those guards here would cause test case failure for those parts.

Comment (stevew817) [reply tests/suites/test_suite_psa_crypto.function:4092]: And since @gilles-peskine-arm beat me to the punch while I was typing my comment: see his answer for the tl;dr.

**Comment by ronald-cron-arm** [comment]
@stevew817 I've added a commit that add macros to check for known alternative implementation limitations and skip the test case if this is the case. I've also re-based as in the definition of the macros I use something that has been merge recently. Please have a look.

**Comment by stevew817** [comment]
@ronald-cron-arm your changes look appropriate to me. I pushed some language/verbiage changes. If you agree, please add your approval.

**Review by ronald-cron-arm — APPROVED** [review]
I am happy with the changes in the documentation of the SKIP_IF macros, thanks for that. I am approving this PR but the part that I've written. For that part, basically the commit "tests: psa: Add macros to skip a test case", please approve Steven if you are ok with it. @tom-daubney-arm please have another look and we need your approval for all the changes.

**Review by tom-daubney-arm — APPROVED** [review]
I have been following along in parallel and agree that all these changes look sensible. Thanks.

**Comment by stevew817** [comment]
@ronald-cron-arm I approve of your commit. I can't explicitly approve through the Github system since I'm the PR author. Marking this ready for merge, then.

**Pull request merged**
